agent/eventstream/eventstream_test.go

```diff
@@ -13,115 +13,86 @@
 	eventStream.StartListening()
 
 	err := eventStream.WriteToEventStream(struct{}{})
-	if err != nil {
-		t.Errorf("Write to event stream failed, err: %v", err)
-	}
-
-	time.Sleep(1 * time.Second)
-
-	cancel()
-	if !listener.called {
-		t.Error("Listener was not invoked")
-	}
+	assert.NoError(t,err)
+	waiter.Wait()
 }
 
 // TestUnsubscribe tests the listener unsubscribed from the
 // event steam will not be notified
 func TestUnsubscribe(t *testing.T) {
-	listener1 := eventListener{called: false}
-	listener2 := eventListener{called: false}
+	waiter1, listener1 := setup()
+	waiter2, listener2 := setup()
 
 	ctx, cancel := context.WithCancel(context.Background())
-	eventStream := NewEventStream("TestUnsubscribe", ctx)
-
-	eventStream.Subscribe("listener1", listener1.recordCall)
-	eventStream.Subscribe("listener2", listener2.recordCall)
+	defer cancel()
 
+	eventStream := NewEventStream("TestUnsubscribe", ctx)
+	eventStream.Subscribe("listener1", listener1)
+	eventStream.Subscribe("listener2", listener2)
 	eventStream.StartListening()
 
+	waiter1.Add(1)
+	waiter2.Add(1)
 	err := eventStream.WriteToEventStream(struct{}{})
-	if err != nil {
-		t.Errorf("Write to event stream failed, err: %v", err)
-	}
+	assert.NoError(t,err)
+	waiter1.Wait()
+	waiter2.Wait()
 
-	time.Sleep(1 * time.Second)
-	if !listener1.called {
-		t.Error("Listener 1 was not invoked")
-	}
-	if !listener2.called {
-		t.Error("Listener 2 was not invoked")
-	}
-
-	listener1.called = false
-	listener2.called = false
 	eventStream.Unsubscribe("listener1")
 
+	waiter2.Add(1)
 	err = eventStream.WriteToEventStream(struct{}{})
-	if err != nil {
-		t.Errorf("Write to event stream failed, err: %v", err)
-	}
-
-	// wait for the event stream to broadcast
-	time.Sleep(1 * time.Second)
-
-	if listener1.called {
-		t.Error("Unsubscribed handler shouldn't be called")
-	}
+	assert.NoError(t,err)
 
-	if !listener2.called {
-		t.Error("Listener 2 was not invoked without unsubscribing")
-	}
-	cancel()
+	waiter1.Wait()
+	waiter2.Wait()
 }
 
 // TestCancelEventStream tests the event stream can
 // be closed by context
 func TestCancelEventStream(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	eventStream := NewEventStream("TestCancelEventStream", ctx)
+	_, listener := setup()
 
-	listener := eventListener{called: false}
-
-	eventStream.Subscribe("listener", listener.recordCall)
-
+	eventStream.Subscribe("listener", listener)
 	eventStream.StartListening()
 	cancel()
 
-	// wait for the event stream to handle cancel
 	time.Sleep(1 * time.Second)
 
 	err := eventStream.WriteToEventStream(struct{}{})
-	if err == nil {
-		t.Error("Write to closed event stream should return an error")
-	}
-	if listener.called {
-		t.Error("Cancelled events context, handler should not be called")
+	assert.Error(t,err)
+}
+
+// setup creates a waitgroup and a function that decrements a WaitGroup when called
+func setup() (*sync.WaitGroup, func(...interface{}) error) {
```

Comment:
Nifty little utility! If you break off any of the others functions identified elsewhere into a test utils package, this is definitely a candidate as it could be put to use in other contexts! Can you give it a more meaningful name either way?